graylog2-server/src/main/java/org/graylog2/users/UserImpl.java

```diff
@@ -144,11 +153,53 @@ public boolean isReadOnly() {
         return Collections.emptyMap();
     }
 
+    @Override
+    public String getFirstName() {
+        final Object firstName = fields.get(FIRST_NAME);
+        return firstName != null ? String.valueOf(firstName) : null;
+    }
+
+    @Override
+    public String getLastName() {
+        final Object lastName = fields.get(LAST_NAME);
+        return lastName != null ? String.valueOf(lastName) : null;
+    }
+
+    /**
+     * @return The full user name formatted as "First Last".
+     * first_name and last_name fields were added in Graylog 4.1.
+     * So, they might not be present for users that existed before.
+     * Fall back to full_name when they are not present.
+     */
     @Override
     public String getFullName() {
+        final Object firstName = fields.get(FIRST_NAME);
+        final Object lastName = fields.get(LAST_NAME);
+        if (firstName != null && lastName != null ) {
+            return String.format(Locale.ENGLISH, "%s %s", firstName, lastName);
+        }
+
         return String.valueOf(fields.get(FULL_NAME));
     }
 
+    /**
+     * Set the user's fullName, firstName, and lastName. The fullName is composed with the {@link #FULL_NAME_FORMAT}
+     * @param firstName The user's first name.
+     * @param lastName The user's last name.
+     */
+    @Override
+    public void setFullName(final String firstName, final String lastName) {
+        Preconditions.checkArgument(StringUtils.isNotBlank(firstName), "A first_name value is required.");
+        Preconditions.checkArgument(StringUtils.isNotBlank(lastName), "A last_name value is required.");
+        fields.put(FIRST_NAME, firstName);
+        fields.put(LAST_NAME, lastName);
+        fields.put(FULL_NAME, String.format(Locale.ENGLISH, FULL_NAME_FORMAT, firstName, lastName));
```

Comment:
We've discussed this and the reason is, that we need this for sorting, searching and paginating which is done in MongoDB.

I think it would be nice to document this here.

Reply:
What also might work is to override `getFields` and calculate fullName at the last moment before it gets stored. https://github.com/Graylog2/graylog2-server/blob/f35df42e165ac570b8b27de3f8eeac85e74ed610/graylog2-server/src/main/java/org/graylog2/plugin/database/Persisted.java#L29